proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+# Native lock
+
+## Summary
+[summary]: #summary
+
+Special-case how `System.Threading.Lock` interacts with the `lock` keyword (calling its `EnterLockScope` method under the hood).
+Add static analysis warnings to prevent accidental misuse of the type where possible.
+
+## Motivation
+[motivation]: #motivation
+
+.NET 9 is introducing a new [`System.Threading.Lock` type](https://github.com/dotnet/runtime/issues/34812)
+as a better alternative to existing monitor-based locking.
+The presence of the `lock` keyword in C# might lead developers to think they can use it with this new type.
+Doing so wouldn't lock according to the semantics of this type but would instead treat it as any other object and would use monitor-based locking.
+
+## Detailed design
+[design]: #detailed-design
+
+Semantics of the lock statement ([§13.13](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/statements.md#1313-the-lock-statement))
+are changed to special-case the `System.Threading.Lock` type:
+
+> A `lock` statement of the form `lock (x) { ... }`
+>
+> 1. **where `x` is an expression of type `System.Threading.Lock`, is precisely equivalent to:**
+>    ```cs
+>    using (x.EnterLockScope())
+>    {
+>        ...
+>    }
+>    ```
+>    **and `System.Threading.Lock` must have the following shape:**
+>    ```cs
+>    namespace System.Threading
+>    {
+>        class Lock
+>        {
+>            public Scope EnterLockScope();
+>    
+>            public ref struct Scope
```

Comment:
does this mean one of thse locks can't be used across an async boundary point?  Is that an intentional/desirable restriction?  

Also, if it's not used across an async boundary point, would the compiler still reject this in a synchronous portion of an async method?  or would ew special case this?

Reply:
Not sure what you mean by "one of these locks". There's currently only one `Lock` and that shouldn't be used across async boundaries, that's right. See also its API proposal: https://github.com/dotnet/runtime/issues/34812

> Also, if it's not used across an async boundary point, would the compiler still reject this in a synchronous portion of an async method? or would ew special case this?

It behaves like the equivalent `using` would, so that's not allowed

Reply:
> Not sure what you mean by "one of these locks".

until there is a requirement that this is sealed, you could have several different suclasses.

Also, i simply mean "an instance of this type in some piece of code".

Reply:
> It behaves like the equivalent using would, so that's not allowed

seems highly unfortunate :)  i may have an async method that still needs to safely lock over some data in a synchronous portion fo it.

Reply:
Hm, if we lower to using, we could probably check/assume that Lock.Scope is safe to be hoisted (like it doesn't contain ref fields) and allow that. Or we could lower to try/catch instead. Or runtime folks could define Lock.Scope as normal non-ref struct.

Reply:
`Lock.Scope` currently uses a ref struct to avoid a redundant thread-local lookup on the exit path, which can improve the performance of several enter/exit use cases by ~5-15%. Maybe the restriction on use of ref structs in async methods could be relaxed in the future such that they could be used, just not across awaits.

Reply:
> seems highly unfortunate :) i may have an async method that still needs to safely lock over some data in a synchronous portion fo it.

It's a bit unfortunate, but for now those could still be refactored into a non-async method.